kirkwood: add support for NETGEAR ReadyNAS Duo v2

CHKDSK88 opened the pull request:
NETGEAR ReadyNAS Duo v2 is a NAS based on Marvell kirkwood SoC.

Specification:
 - Processor Marvell 88F6282 (1.6 GHz)
 - 256MB RAM
 - 128MB NAND
 - 1x GBE LAN port (PHY: Marvell 88E1318)
 - 1x USB 2.0
 - 2x USB 3.0
 - 2x SATA
 - 3x button
 - 5x leds
 - serial on J5 connector accessible from rear panel
   (115200 8N1) (VCC,TX,RX,GND) (3V3 LOGIC!)

Installation by USB + serial:
  - Copy initramfs image to fat32 usb drive
  - Connect pendrive to USB 2.0 front socket
  - Connect serial console
  - Stop booting in u-boot
  - Do:
	usb reset
	fatload usb 0:1 0x1200000 openwrt-kirkwood-netgear_readynas-duo-v2-initramfs-uImage
	setenv bootargs 'console=ttyS0,115200n8 earlyprintk'
	bootm 0x1200000
  - copy sysupgrade image via ssh.
  - run sysupgrade

Installation by TFTP + serial:
  - Setup TFTP server and copy initramfs image
  - Connect serial console
  - Stop booting in u-boot
  - Do:
	setenv serverip 192.168.1.1
	setenv ipaddr 192.168.1.2
	setenv bootargs 'console=ttyS0,115200n8 earlyprintk'
	tftpboot 0x1200000 openwrt-kirkwood-netgear_readynas-duo-v2-initramfs-uImage
	bootm 0x1200000
  - copy sysupgrade image via ssh.
  - run sysupgrade

Known issues:
  - Power button and PHY INTn pin are connected to the same GPIO. It
    causes that every network restart button is pressed in system.
    As workaround, button is used as regular BTN_1.

For more info please look at file:
RND_5.3.13_WW.src/u-boot/board/mv_feroceon/mv_hal/usibootup/usibootup.c
from Netgear GPL sources.

Tested-by: Raylynn Knight <rayknight@me.com>
Tested-by: Lech Perczak <lech.perczak@gmail.com>
Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

CHKDSK88 committed:
kirkwood: add support for NETGEAR ReadyNAS Duo v2

NETGEAR ReadyNAS Duo v2 is a NAS based on Marvell kirkwood SoC.

Specification:
 - Processor Marvell 88F6282 (1.6 GHz)
 - 256MB RAM
 - 128MB NAND
 - 1x GBE LAN port (PHY: Marvell 88E1318)
 - 1x USB 2.0
 - 2x USB 3.0
 - 2x SATA
 - 3x button
 - 5x leds
 - serial on J5 connector accessible from rear panel
   (115200 8N1) (VCC,TX,RX,GND) (3V3 LOGIC!)

Installation by USB + serial:
  - Copy initramfs image to fat32 usb drive
  - Connect pendrive to USB 2.0 front socket
  - Connect serial console
  - Stop booting in u-boot
  - Do:
	usb reset
        setenv bootargs 'console=ttyS0,115200n8 earlyprintk'
        setenv bootcmd 'nand read.e 0x1200000 0x200000 0x600000;bootm 0x1200000'
        saveenv
	fatload usb 0:1 0x1200000 openwrt-kirkwood-netgear_readynas-duo-v2-initramfs-uImage
	bootm 0x1200000
  - copy sysupgrade image via ssh.
  - run sysupgrade

Installation by TFTP + serial:
  - Setup TFTP server and copy initramfs image
  - Connect serial console
  - Stop booting in u-boot
  - Do:
	setenv bootargs 'console=ttyS0,115200n8 earlyprintk'
	setenv bootcmd 'nand read.e 0x1200000 0x200000 0x600000;bootm 0x1200000'
	saveenv
	setenv serverip 192.168.1.1
	setenv ipaddr 192.168.1.2
	tftpboot 0x1200000 openwrt-kirkwood-netgear_readynas-duo-v2-initramfs-uImage
	bootm 0x1200000
  - copy sysupgrade image via ssh.
  - run sysupgrade

Known issues:
  - Power button and PHY INTn pin are connected to the same GPIO. It
    causes that every network restart button is pressed in system.
    As workaround, button is used as regular BTN_1.

For more info please look at file:
RND_5.3.13_WW.src/u-boot/board/mv_feroceon/mv_hal/usibootup/usibootup.c
from Netgear GPL sources.

Tested-by: Raylynn Knight <rayknight@me.com>
Tested-by: Lech Perczak <lech.perczak@gmail.com>
Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

CHKDSK88 committed:
packages: kernel: add i2c hwmon g762 kmod package

This patch adds kernel module for Global Mixed-mode Technology Inc
G762 and G763 fan speed PWM controller chips.

Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

adschm commented:
So, this can only be flashed with serial access?

CHKDSK88 commented:
@adschm  
> So, this can only be flashed with serial access?

Yes, serial is required, but without disassembling the case. It's on rear panel.

RaylynnKnight commented:
This will need clarification as to exactly what Netgear ReadyNAS v2 this is for.  There are multiple Netgear ReadyNAS v2 units two of which contain a Kirkwood SOC.  The Netgear Model RND-4B marketing name [Netgear ReadyNAS NV+ v2](https://wikidevi.wi-cat.ru/Netgear_ReadyNAS_NV%2B_RND4000) is a 4-bay NAS, The Netgear Model RND-2B marketing name [Netgear ReadyNAS Duo v2](https://wikidevi.wi-cat.ru/Netgear_ReadyNAS_Duo_RND2000) is a 2-bay NAS.  Based on your description this should be Netgear ReadyNAS Duo v2 to avoid confusion.

CHKDSK88 commented:
@RaylynnKnight 

Thank You. I added "Duo" to device name.

RaylynnKnight commented:
Has this been tested on an actual device?  I pulled the git repository branch for CHKDSK88:netgear-readynas-v2.  Did a build resulting in:

drwxr-xr-x 3 rayk rayk     4096 Sep  5 04:28 .
drwxr-xr-x 3 rayk rayk     4096 Sep  5 04:09 ..
-rw-r--r-- 1 rayk rayk     2452 Sep  5 04:09 config.buildinfo
-rw-r--r-- 1 rayk rayk      337 Sep  5 04:09 feeds.buildinfo
-rw-r--r-- 1 rayk rayk 11448926 Sep  5 04:28 openwrt-kirkwood-netgear_readynas-duo-v2-initramfs-uImage
-rw-r--r-- 1 rayk rayk     4220 Sep  5 04:28 openwrt-kirkwood-netgear_readynas-duo-v2.manifest
-rw-r--r-- 1 rayk rayk 12585771 Sep  5 04:28 openwrt-kirkwood-netgear_readynas-duo-v2-squashfs-sysupgrade.bin
drwxr-xr-x 2 rayk rayk     4096 Sep  5 04:28 packages
-rw-r--r-- 1 rayk rayk      620 Sep  5 04:28 sha256sums
-rw-r--r-- 1 rayk rayk       18 Sep  5 04:09 version.buildinfo

And then followed the "Installation by USB + serial:" instructions with the following result:

11448926 bytes read
Marvell>> bootm 0x1200000
## Booting image at 01200000 ...
   Image Name:   ARM OpenWrt Linux-5.4.61
   Created:      2020-09-04  20:10:09 UTC
   Image Type:   ARM Linux Kernel Image (uncompressed)
   Data Size:    11448862 Bytes = 10.9 MB
   Load Address: 00008000
   Entry Point:  00008000
   Verifying Checksum ... OK

And it has been hung here for over 10 minutes.

CHKDSK88 commented:
@RaylynnKnight 

Could You paste result of u-boot ```printenv```?
Did You try compile unitramfs image with less packages?

CHKDSK88 commented:
```
         __  __                      _ _
        |  \/  | __ _ _ ____   _____| | |
        | |\/| |/ _` | '__\ \ / / _ \ | |
        | |  | | (_| | |   \ V /  __/ | |
        |_|  |_|\__,_|_|    \_/ \___|_|_|
 _   _     ____              _
| | | |   | __ )  ___   ___ | |_ 
| | | |___|  _ \ / _ \ / _ \| __| 
| |_| |___| |_) | (_) | (_) | |_ 
 \___/    |____/ \___/ \___/ \__| 
 ** MARVELL BOARD: DB-88F6282A-BP LE 

U-Boot 1.1.4 (Jun 29 2012 - 16:06:40) Marvell version: 3.4.27
Netgear version: Uboot-1_1_4-NetgearDUOV3-V1009

U-Boot code: 00600000 -> 0067FFF0  BSS: -> 006D0120

Soc: MV88F1155 Rev 1 (DDR3)
CPU running @ 1600Mhz L2 running @ 533Mhz
SysClock = 533Mhz , TClock = 200Mhz 

DRAM unknown CAL  tRP = 8 tRAS = 20 tRCD=8
DRAM CS[0] base 0x00000000   size 256MB 
DRAM Total size 256MB  16bit width
Addresses 8M - 0M are saved for the U-Boot usage.
Mem malloc Initialization (8M - 7M): Done
NAND:128 MB
Flash:  0 kB

CPU : Marvell Feroceon (Rev 1)

Streaming disabled 
Write allocate disabled


USB 0: host mode
PEX 0: PCI Express Root Complex Interface
PEX interface detected Link X1
Switch On !

Net:   egiga0 [PRIME]
Hit any key to stop autoboot:  0 
Marvell>> usb reset
(Re)start USB...
USB:   scanning bus for devices... 2 USB Device(s) found
Waiting for storage device(s) to settle before scanning...
1 Storage Device(s) found
Marvell>> fatload usb 0:1 0x1200000 openwrt-kirkwood-netgear_readynas-duo-v2-initramfs-uImage
reading openwrt-kirkwood-netgear_readynas-duo-v2-initramfs-uImage
.............................................................................................................................................................................................................................................................................................................................................................................................................

4076182 bytes read
Marvell>> bootm 0x1200000
## Booting image at 01200000 ...
   Image Name:   ARM OpenWrt Linux-5.4.61
   Created:      2020-09-04  20:10:09 UTC
   Image Type:   ARM Linux Kernel Image (uncompressed)
   Data Size:    4076118 Bytes =  3.9 MB
   Load Address: 00008000
   Entry Point:  00008000
   Verifying Checksum ... OK
OK

Starting kernel ...

[    0.000000] Booting Linux on physical CPU 0x0
[    0.000000] Linux version 5.4.61 (builder@buildhost) (gcc version 8.4.0 (OpenWrt GCC 8.4.0 r13663+675-5667ccbf16)) #0 Fri Sep 4 20:10:09 2020
[    0.000000] CPU: Feroceon 88FR131 [56251311] revision 1 (ARMv5TE), cr=0005397f
[    0.000000] CPU: VIVT data cache, VIVT instruction cache
[    0.000000] OF: fdt: Machine model: NETGEAR ReadyNAS Duo v2
[    0.000000] Memory policy: Data cache writeback
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 64960
[    0.000000] Kernel command line: console=ttyS0,115200n8 earlyprintk reason=normal mtdparts=nand_mtd:0x180000@0(u-boot),0x20000@0x180000(u-boot-env),0x600000@0x200000(uImage),0x1000000@0x800000(minirootfs),0x6800000@0x1800000(jffs2);
[    0.000000] Dentry cache hash table entries: 32768 (order: 5, 131072 bytes, linear)
[    0.000000] Inode-cache hash table entries: 16384 (order: 4, 65536 bytes, linear)
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
```

CHKDSK88 commented:
@RaylynnKnight 

Did You have time for more tests?

CHKDSK88 commented:
Binary images, if someone would like to test:
[openwrt-kirkwood-netgear_readynas-duo-v2-images.zip](https://github.com/openwrt/openwrt/files/5317845/openwrt-kirkwood-netgear_readynas-duo-v2-images.zip)

RaylynnKnight commented:
You can add a Test By from me.  Installed successfully.

CHKDSK88 commented:
@RaylynnKnight 

Done. Thank You.

adschm commented on target/linux/kirkwood/patches-5.4/112-readynas_duo_v2.patch:
```diff
@@ -0,0 +1,72 @@
+ 
+-		power_led {
++		led_power: power_led {
+ 			label = "status:blue:power_led";
```
Is there a specific reason for these complicated (and non-standard) LED labels? Why not just

blue:power
blue:disk1
blue:disk2

?

adschm replied:
Err, please ignore what I just wrote, I'm too stupid to understand a patch when I see it (and the fact that this is actually upstream). Sorry.

CHKDSK88 commented:
@adschm 

Are You plan to merge it?

RaylynnKnight commented:
@adschm Any chance this could be merged soon?  I have another device (Netgear ReadyNAS NV+ V2) that would extend from the support for this device.

CHKDSK88 commented:
Rebased to current master.

@ynezz 
Any chance for merge it?

bobafetthotmail commented:
@aparcar maybe you can look at this too?

mans0n commented on target/linux/kirkwood/base-files/lib/upgrade/platform.sh:
```diff
@@ -16,6 +16,13 @@ platform_do_upgrade() {
+		mkdir -p /var/lock
+		touch /var/lock/fw_printenv.lock
+		fw_setenv bootargs 'console=ttyS0,115200n8 earlyprintk'
+		fw_setenv bootcmd 'nand read.e 0x1200000 0x200000 0x600000;bootm 0x1200000'
```
Hmm... Can we run these commands at installation time and drop these lines here?

CHKDSK88 replied:
Ok, I will try detect if sysupgrade is called from initramfs and run it only then. But I could test it after 2-3 months, because my device is away from me now.

CHKDSK88 replied:
I must think about it in the next week.

CHKDSK88 replied:
@adschm @mans0n 
Any ideas, how to detect if sysupgrade is running from initramfs image?

mans0n replied:
I'm not sure... But I thought we can just run those commands in U-Boot shell before the tftpboot, am I missing something?

Leo-PL replied:
Or just check if they match already. Either would be fine by me.

Leo-PL replied:
Alternatively, maybe at least the "bootargs" part could be added to device tree as "chosen" property in /aliases in device tree?

CHKDSK88 replied:
I prepared new version: envs are configured only during first installation via initramfs. If someone want override it, changes wouldn't be destroyed via regular sysupgrade.

mans0n commented on target/linux/kirkwood/image/Makefile:
```diff
@@ -136,6 +136,18 @@ define Device/linksys_ea4500
+  DEVICE_DTS := kirkwood-netgear_readynas_duo_v2
+  KERNEL_IN_UBI :=
+  IMAGES := sysupgrade.bin
+  DEVICE_PACKAGES := kmod-ata-marvell-sata kmod-fs-ext4 kmod-gpio-button-hotplug \
```
This line exceeds 80 characters...

mans0n commented on target/linux/kirkwood/base-files/etc/init.d/hwmon_fancontrol:
```diff
@@ -11,6 +11,10 @@ boot() {
 		;;
+	netgear,readynas-duo-v2)
+		path_to_hwmon='/sys/class/hwmon/hwmon0'
+		echo 1200 > "$path_to_hwmon/fan1_target" #set target rpm
```
(I cannot add a comment on a line that is not in the diff, so I'll write this here)

> `# configuring (lm85/lm63) onboard temp/fan controller to run the fan on its own`

Maybe we can add "g762" to `(lm85/lm63)`, or remove it entirely.

And could you add a space after `#`?

mans0n reviewed (commented):
Just some nitpicks.

mans0n commented:
Does `poweroff` command work on this device?
If yes, then maybe we can link the power button to the command.
If not, then isn't it the poweroff procedure itself, not gpio-poweroff, that is buggy?

CHKDSK88 commented:
@mans0n 

> Does `poweroff` command work on this device?

Simple poweroff cause, that it can't be turned on from button and WOL won't work. It need to reset by plug off power supply.

> If yes, then maybe we can link the power button to the command.
> If not, then isn't it the poweroff procedure itself, not gpio-poweroff, that is buggy?

Poweroff in stock firmware cause  setting two bits (PHY register: Page 3 Reg 17 Bit 4 and RTC register: SL2(BIT5) of RS5C372a control 1 register (0xe0)) and reboot. If bootloader detect that state, it enable WOL and suspend device. Then it wait for WOL signal or button.

If gpio-poweroff is used, power button causes state, when is impossible to turn it on, without plug off the power supply.

mans0n commented:
So...you mean that the effect of simple poweroff and gpio-poweroff is totally same?

Asking to be sure, what is the effect of gpio-poweroff pin (gpio30)? Does the stock FW also use it?

danitool commented:
> @mans0n
> 
> > Does `poweroff` command work on this device?
> 
> Simple poweroff cause, that it can't be turned on from button and WOL won't work. It need to reset by plug off power supply.
> 
> > If yes, then maybe we can link the power button to the command.
> > If not, then isn't it the poweroff procedure itself, not gpio-poweroff, that is buggy?
> 
> Poweroff in stock firmware cause setting two bits (PHY register: Page 3 Reg 17 Bit 4 and RTC register: SL2(BIT5) of RS5C372a control 1 register (0xe0)) and reboot. If bootloader detect that state, it enable WOL and suspend device. Then it wait for WOL signal or button.
> 
> If gpio-poweroff is used, power button causes state, when is impossible to turn it on, without plug off the power supply.

Hi @CHKDSK88, your NAS is using a marvell PHY right?, then you may need the same poweroff-restart driver I made for the LinkStation LS421DE NAS (see https://github.com/openwrt/openwrt/commit/59c200ca0adec1759f09b129424361). It uses the INTn at the PHY to wake up the machine, restart or power off depending on its state.

However, the PHY registers you point to don't make much sense to me. 

About the RTC, probably the register has to do with the wakealarm pin interrupt, hopefully it is already cleared if a wake alarm is scheduled or not.

For testing the Linkstation poweroff-restart driver you need to add your board to the compatible list in the driver itself.

Regards
Daniel

CHKDSK88 commented:
@mans0n  
> 
> So...you mean that the effect of simple poweroff and gpio-poweroff is totally same?

gpio-poweroff does nothing special. It should reboot device, but it halt.
 
> Asking to be sure, what is the effect of gpio-poweroff pin (gpio30)? Does the stock FW also use it?

I can't find any info about this pin in GPL sources.

@danitool 
In Netgear it's different. It use "ACPOWER fail recorder using RS5C372A one bit." If device should be powered off, it change bit and reboot. After reboot (or cold start) uboot checks value of this register and if it's configured, it allow to boot. In other case, enable WOL, suspend cpu and wait for button.

mans0n commented:
Sorry, but I still don't get it. If the poweroff pin does not turn off the board power, and the stock FW does not use it, and it does not work for us, then why did you use it at the first place? Where did you get the pin number?

CHKDSK88 commented:
Sorry, I was unclear.

After "shutdown" command it will be halted. But in this state it's impossible to use WOL or power button for turn on again. It should be restarted and halted by u-boot with proper preparation. After that, it's possible to use button and WOL for turn on.

mans0n commented:
So if I understood correctly,

1. In the current state, `shutdown` command and power button does not have any difference in effect. Both halt the machine but then it's impossible to boot again without a power cycle.
2. On stock FW, power button and reboot button does the same, i.e. warm-reboot the device, but the former sets some regs before the reboot, and the bootloader detects it and suspends the device.
3. There is no way to "power off" the device. gpio-poweroff pin (gpio30) does not turn off the board power, it does not reset the CPU, it does nothing.

Then,

4. The OpenWrt poweroff is not broken, it's just not enough for the device. Anyway, you'll get the same result if you shutdown a common router device which has no board power control. (It will halt and it'll need a power cycle to reboot)
5. If we somehow have a way to access those registers in userspace (e.g. via custom driver or `devmem` command), we can write a `/etc/rc.button` script and do the reg stuff before poweroff, thereby fixing the power button.
6. We can drop the gpio-poweroff node as it is useless.

Leo-PL commented:
I got my hands on one device in hope to get something more supportable, than my old Synology DS211j. I think, that a proper poweroff driver should be developed, as it was done for Qnap and Synology devices - when the device arrives, I'll cobble something up.

Leo-PL commented:
@CHKDSK88 I got my device today, so I can do some testing for you, so proper support could be merged quicker.

Leo-PL commented:
Also, I did some digging around the factory image. It looks like it's a tarball containing kernel, U-boot and rootfs images, prefixed with a header:
```
$ binwalk RAIDiator-arm-5.3.13 

DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
16384         0x4000          POSIX tar archive (GNU), owner user name: "5"

$ dd if=RAIDiator-arm-5.3.13 bs=16k skip=1 | tar tv
-rw-r--r-- root/root       390 2017-05-24 06:36 csums.md5
-rw-r--r-- root/root   4382897 2017-05-24 06:36 initrd-1.0.66
-rw-r--r-- root/root   3442424 2017-05-24 06:36 kernel-DUO-1.0.23
-rw-r--r-- root/root   3440516 2017-05-24 06:36 kernel-NV+-1.0.26
-rw-r--r-- root/root  74506696 2017-05-24 06:18 root.tlz
-rw-r--r-- root/root    941140 2017-05-24 06:36 supplmnt.tlz
-rw-r--r-- root/root    475340 2017-05-24 06:36 uboot-DUO-1.0.06
-rw-r--r-- root/root    476444 2017-05-24 06:36 uboot-NV+-1.0.05
-rw-r--r-- root/root        42 2017-05-24 06:36 version.txt
```

CHKDSK88 committed:
kernel: generic: backport linkstation-poweroff driver

Linkstation poweroff driver was added to mvebu target, but is required
for kirkwood target too.
This commit make two changes:
- move linkstation-poweroff support patch from mvebu to generic and
replace upstream accepted version
- backport small linkstation-poweroff fix from 5.12

Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

CHKDSK88 commented:
I rebased and refreshed PR. I use @danitool driver and poweroff work now perfectly. 

Last thing is to run fw_setenv only if image is flashed from initramfs. I plan to work with this at monday. And it will be ready.

RaylynnKnight commented:
Could you use fw_getenv to retrieve the values and only set them if they don’t match?

Ray



> On Jun 18, 2021, at 11:25 AM, Paweł Dembicki ***@***.***> wrote:
> 
> 
> @CHKDSK88 commented on this pull request.
> 
> In target/linux/kirkwood/base-files/lib/upgrade/platform.sh:
> 
> > @@ -16,6 +16,13 @@ platform_do_upgrade() {
>  	linksys,ea4500)
>  		platform_do_upgrade_linksys "$1"
>  		;;
> +	netgear,readynas-duo-v2)
> +		mkdir -p /var/lock
> +		touch /var/lock/fw_printenv.lock
> +		fw_setenv bootargs 'console=ttyS0,115200n8 earlyprintk'
> +		fw_setenv bootcmd 'nand read.e 0x1200000 0x200000 0x600000;bootm 0x1200000'
> 
> @adschm @mans0n
> Any ideas, how to detect if sysupgrade is running from initramfs image?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

emesio commented:
Hi friends. Any chances for new bins? 
Happy to see this topic is alive. Got a ReadyNasDuo and can help with any tests if needed (but I'm more like user, not advanced linux admin ;). On old bins OpenWRT works nice, but theres old kernel, some packages incompability, and so on (and this poweroff issue, which looks like solved from new comments). 
Good job!

CHKDSK88 commented:
@Leo-PL @RaylynnKnight 

I made corrections. It's ready.
Please look if everything works ok. 

I plan to send driver patch upstream, if You confirm, that it works on Your devices.

Leo-PL commented:
Everything works for me, power-down and WoL included. No need to send a very detailed test report here, as I did for NSA310s. Great work!
I even went so far, as to backport your changes to 19.07 to get the device running on stable release, but there the linkstation-poweroff crashed, most likely due to my oversight.

CHKDSK88 commented:
@Leo-PL  
> 
> Everything works for me, power-down and WoL included. No need to send a very detailed test report here, as I did for NSA310s. Great work!
> Tested-By: Lech Perczak [lech.perczak@gmail.com](mailto:lech.perczak@gmail.com)
> 

Thank You for testing...

> I even went so far, as to backport your changes to 19.07 to get the device running on stable release, but there the linkstation-poweroff crashed, most likely due to my oversight.
> 
> Regarding the driver, it might need a bit of refactoring, otherwise the kernel folks may complain - probably about using static struct phy_device and struct power_off_cfg, as well tying the compatible string directly to machine. Please take a look, for example, at qnap-poweroff driver for an example, how to write such driver - the changes should be pretty straightforward, as from logical point of view - current driver is fine.

And for taking the look into driver code. I'm not a kernel specialist, so feedback is important for me.
I think about those compatible strings. Linkstation-poweroff is very fresh driver and author used the compatible string directly to machine.  So I adjust existing code, without any big changes. 
New thing is only `power_off_cfg`, but idea about this struct was taken from qnap-poweroff.c.

Leo-PL commented:
I see that only the patches were renamed without changing code itself, so still +1 from me. Regarding the driver, since it's upstreamed already, maybe changes to compatible strings, etc, won't be necessary. I think you might have a go at it at LKML, but please do, at very least, a compile testtest at current master. And do link discussion here, so our community could participate in review as well. If you wish, I can do a more in-depth review of the driver for you beforehand.

CHKDSK88 commented:
@Leo-PL Thank You for feedback. I ran checkpatch and compile test against upstream master. Then I sent it to ML:

https://marc.info/?l=linux-pm&m=162452630626342&w=2

I hope, that this PR will be merged soon.

CHKDSK88 commented:
@adschm 
Any chance for merge?

Leo-PL commented:
I just rebased this PR together with my patchset, and for some strange reason, when OpenWrt boots, at the time around link establishes, a poweroff is triggered by kernel, and system turns off. This did not happen during my previous tests, and really nothing has changed in this PR, so I don't suspect it to have the problem. 

I really, really wonder what happened - or what did I do wrong, since the device tree hasn't changed, and the same exact thing happens at master, and with a backport to 19.07 I did myself (not really complete, due to linkstation-poweroff driver crashing at poweroff). A regression in two stable kernel lines? 
Now trying a clean build from scratch...

CHKDSK88 commented:
@Leo-PL 

> Also, freshly build initramfs images for some reason, seem to crash very early on, even before the initial "OK" is printed. Can they be too big? With current master, initramfs-uImage weighs in 12337598 bytes for me.

Initramfs image must be small. I'm not sure what size of image is maximum for proper loading. I had the same problem when I test images prepared by @obsy from eko.one.pl.


> I just rebased this PR together with my patchset, and for some strange reason, when OpenWrt boots, at the time around link establishes, a poweroff is triggered by kernel, and system turns off. This did not happen during my previous tests, and really nothing has changed in this PR, so I don't suspect it to have the problem.

I have the same problem.

> One thought came to my mind. I don't have the schematics for the device - but isn't WoL working by asserting the "power button" line internally?

Exactly, it looks that is problem with LED2/INTn pin connected to power button. It could pull down power button.
I pushed partial solution of powering-off problem. Now sometimes work perfect, sometimes not. It depends, how fast phy will be configured. If gpio-hotplug is faster, then it reboot. If phy are faster, it works. 

The only way is to delay start of gpio-hotplug or do some changes in phy configuration. Do You have any idea how to make it better?

Leo-PL commented:
Thanks for workaround. I intend to try to bisect the kernel to find the root cause of the problem, as it sounds very much like stable kernel regression. In my day job, recently I've stumbled upon other issues related to libphy as well.

CHKDSK88 commented:
Please look how many ```marvell,reg-init``` is called. Every call causes change state of power button. 

```
[    0.000000] Booting Linux on physical CPU 0x0
[    0.000000] Linux version 5.4.132 (krolik@tempest2) (gcc version 8.4.0 (OpenWrt GCC 8.4.0 r0+16244-75f9daa8fc)) #0 Fri Jul 23 13:59:21 2021
[    0.000000] CPU: Feroceon 88FR131 [56251311] revision 1 (ARMv5TE), cr=0005397f
[    0.000000] CPU: VIVT data cache, VIVT instruction cache
[    0.000000] OF: fdt: Machine model: NETGEAR ReadyNAS Duo v2
[    0.000000] Memory policy: Data cache writeback
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 64960
[    0.000000] Kernel command line: console=ttyS0,115200n8 earlyprintk
[    0.000000] Dentry cache hash table entries: 32768 (order: 5, 131072 bytes, linear)
[    0.000000] Inode-cache hash table entries: 16384 (order: 4, 65536 bytes, linear)
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory: 243784K/262144K available (5130K kernel code, 173K rwdata, 1380K rodata, 8192K init, 205K bss, 18360K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 16, nr_irqs: 16, preallocated irqs: 16
[    0.000000] random: get_random_bytes called from start_kernel+0x2d4/0x4fc with crng_init=0
[    0.000000] clocksource: orion_clocksource: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 9556302233 ns
[    0.000006] sched_clock: 32 bits at 200MHz, resolution 5ns, wraps every 10737418237ns
[    0.000026] Switching to timer-based delay loop, resolution 5ns
[    0.000085] Calibrating delay loop (skipped), value calculated using timer frequency.. 400.00 BogoMIPS (lpj=2000000)
[    0.000099] pid_max: default: 32768 minimum: 301
[    0.000254] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.000270] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.000958] CPU: Testing write buffer coherency: ok
[    0.001644] Setting up static identity map for 0x100000 - 0x10003c
[    0.001830] mvebu-soc-id: MVEBU SoC ID=0x6282, Rev=0x1
[    0.004459] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.004477] futex hash table entries: 256 (order: -1, 3072 bytes, linear)
[    0.004576] pinctrl core: initialized pinctrl subsystem
[    0.005337] NET: Registered protocol family 16
[    0.005615] DMA: preallocated 256 KiB pool for atomic coherent allocations
[    0.006373] cpuidle: using governor ladder
[    0.006632] Feroceon L2: Enabling L2
[    0.006664] Feroceon L2: Cache support initialised.
[    0.009728] No ATAGs?
[    0.024058] SCSI subsystem initialized
[    0.024736] usbcore: registered new interface driver usbfs
[    0.024777] usbcore: registered new interface driver hub
[    0.024812] usbcore: registered new device driver usb
[    0.024933] workqueue: max_active 576 requested for napi_workq is out of range, clamping between 1 and 512
[    0.026793] clocksource: Switched to clocksource orion_clocksource
[    0.027297] thermal_sys: Registered thermal governor 'step_wise'
[    0.027465] NET: Registered protocol family 2
[    0.027567] IP idents hash table entries: 4096 (order: 3, 32768 bytes, linear)
[    0.027984] tcp_listen_portaddr_hash hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.028011] TCP established hash table entries: 2048 (order: 1, 8192 bytes, linear)
[    0.028040] TCP bind hash table entries: 2048 (order: 1, 8192 bytes, linear)
[    0.028066] TCP: Hash tables configured (established 2048 bind 2048)
[    0.028130] UDP hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.028150] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.028288] NET: Registered protocol family 1
[    0.028311] PCI: CLS 0 bytes, default 32
[    0.103810] workingset: timestamp_bits=14 max_order=16 bucket_order=2
[    0.107872] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.107881] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.115851] kirkwood-pinctrl f1010000.pin-controller: registered pinctrl driver
[    0.116552] mvebu-gpio f1010140.gpio: IRQ index 3 not found
[    0.116899] mvebu-pcie mbus@f1000000:pcie@82000000: PCI host bridge to bus 0000:00
[    0.116913] pci_bus 0000:00: root bus resource [bus 00-ff]
[    0.116921] pci_bus 0000:00: root bus resource [mem 0xe0000000-0xefffffff]
[    0.116929] pci_bus 0000:00: root bus resource [io  0x1000-0xeffff]
[    0.117004] pci 0000:00:01.0: [11ab:6282] type 01 class 0x060400
[    0.117029] pci 0000:00:01.0: reg 0x38: [mem 0x00000000-0x000007ff pref]
[    0.117925] PCI: bus0: Fast back to back transfers disabled
[    0.117937] pci 0000:00:01.0: bridge configuration invalid ([bus 00-00]), reconfiguring
[    0.118036] pci 0000:01:00.0: [1033:0194] type 00 class 0x0c0330
[    0.118079] pci 0000:01:00.0: reg 0x10: [mem 0x40000000-0x40001fff 64bit]
[    0.118212] pci 0000:01:00.0: PME# supported from D0 D3hot D3cold
[    0.118245] pci 0000:01:00.0: 2.000 Gb/s available PCIe bandwidth, limited by 2.5 GT/s x1 link at 0000:00:01.0 (capable of 4.000 Gb/s with 5 GT/s x1 link)
[    0.119125] PCI: bus1: Fast back to back transfers disabled
[    0.119137] pci_bus 0000:01: busn_res: [bus 01-ff] end is updated to 01
[    0.119276] pci 0000:00:01.0: BAR 8: assigned [mem 0xe0000000-0xe00fffff]
[    0.119286] pci 0000:00:01.0: BAR 6: assigned [mem 0xe0100000-0xe01007ff pref]
[    0.119297] pci 0000:01:00.0: BAR 0: assigned [mem 0xe0000000-0xe0001fff 64bit]
[    0.119314] pci 0000:00:01.0: PCI bridge to [bus 01]
[    0.119324] pci 0000:00:01.0:   bridge window [mem 0xe0000000-0xe00fffff]
[    0.119655] Serial: 8250/16550 driver, 2 ports, IRQ sharing disabled
[    0.120124] printk: console [ttyS0] disabled
[    0.120178] f1012000.serial: ttyS0 at MMIO 0xf1012000 (irq = 26, base_baud = 12500000) is a 16550A
[    0.639937] printk: console [ttyS0] enabled
[    0.645010] nand: device found, Manufacturer ID: 0xad, Chip ID: 0xf1
[    0.651410] nand: Hynix H27U1G8F2BTR-BC
[    0.655260] nand: 128 MiB, SLC, erase size: 128 KiB, page size: 2048, OOB size: 64
[    0.662876] Scanning device for bad blocks
[    0.743770] 4 fixed-partitions partitions found on MTD device orion_nand
[    0.750509] Creating 4 MTD partitions on "orion_nand":
[    0.755673] 0x000000000000-0x000000180000 : "u-boot"
[    0.761166] 0x000000180000-0x0000001a0000 : "u-boot-env"
[    0.766949] 0x000000200000-0x000000800000 : "kernel"
[    0.772361] 0x000000800000-0x000008000000 : "ubi"
[    0.778115] libphy: Fixed MDIO Bus: probed
[    0.782769] libphy: orion_mdio_bus: probed
[    0.799396] mv643xx_eth: MV-643xx 10/100/1000 ethernet driver version 1.4
[    0.807661] Marvell 88E1318S f1072004.mdio-bus-mii:00: marvell,reg-init!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
[    0.820348] mv643xx_eth_port mv643xx_eth_port.0 eth0: port 0 with MAC address 2c:b0:5d:bf:0d:d4
[    1.846798] rtc-mv f1010300.rtc: internal RTC not ticking
[    1.852290] i2c /dev entries driver
[    1.856764] watchdog: f1020300.watchdog-timer: driver supplied timeout (4294967295) out of range
[    1.865610] watchdog: f1020300.watchdog-timer: falling back to default timeout (21)
[    1.873420] orion_wdt: Initial timeout 21 sec
[    1.948597] marvell-cesa f1030000.crypto: CESA device successfully registered
[    1.956409] NET: Registered protocol family 10
[    1.962138] Segment Routing with IPv6
[    1.965870] NET: Registered protocol family 17
[    1.970486] 8021q: 802.1Q VLAN Support v1.8
[    1.978046] UBI: auto-attach mtd3
[    1.981384] ubi0: attaching mtd3
[    2.126327] ubi0: scanning is finished
[    2.140433] ubi0: attached mtd3 (name "ubi", size 120 MiB)
[    2.145943] ubi0: PEB size: 131072 bytes (128 KiB), LEB size: 129024 bytes
[    2.152865] ubi0: min./max. I/O unit sizes: 2048/2048, sub-page size 512
[    2.159605] ubi0: VID header offset: 512 (aligned 512), data offset: 2048
[    2.166416] ubi0: good PEBs: 960, bad PEBs: 0, corrupted PEBs: 0
[    2.172452] ubi0: user volume: 2, internal volumes: 1, max. volumes count: 128
[    2.179708] ubi0: max/mean erase counter: 4/1, WL threshold: 4096, image sequence number: 1927344411
[    2.188888] ubi0: available PEBs: 0, total reserved PEBs: 960, PEBs reserved for bad PEB handling: 20
[    2.198670] ubi0: background thread "ubi_bgt0d" started, PID 463
[    2.205963] block ubiblock0_0: created from ubi0:0(rootfs)
[    2.211502] ubiblock: device ubiblock0_0 (rootfs) set to be root filesystem
[    2.218510] hctosys: unable to open rtc device (rtc0)
[    2.247188] Freeing unused kernel memory: 8192K
[    2.251774] Run /init as init process
[    2.376819] random: fast init done
[    2.447777] init: Console is alive
[    2.451304] init: - watchdog -
[    2.462403] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    2.487861] ehci_hcd: USB 2.0 'Enhanced' Host Controller (EHCI) Driver
[    2.497016] ehci-fsl: Freescale EHCI Host controller driver
[    2.504897] ehci-orion: EHCI orion driver
[    2.509084] orion-ehci f1050000.ehci: EHCI Host Controller
[    2.514611] orion-ehci f1050000.ehci: new USB bus registered, assigned bus number 1
[    2.522429] orion-ehci f1050000.ehci: irq 30, io mem 0xf1050000
[    2.556812] orion-ehci f1050000.ehci: USB 2.0 started, EHCI 1.00
[    2.563307] hub 1-0:1.0: USB hub found
[    2.567246] hub 1-0:1.0: 1 port detected
[    2.573823] ehci-platform: EHCI generic platform driver
[    2.594540] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    2.604537] init: - preinit -
[    2.729846] random: jshn: uninitialized urandom read (4 bytes read)
[    2.763459] random: jshn: uninitialized urandom read (4 bytes read)
[    2.783925] random: jshn: uninitialized urandom read (4 bytes read)
[    2.898879] Marvell 88E1318S f1072004.mdio-bus-mii:00: marvell,reg-init!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
[    2.913028] mv643xx_eth_port mv643xx_eth_port.0 eth0: link up, 1000 Mb/s, full duplex, flow control disabled
[    2.922915] mv643xx_eth_port mv643xx_eth_port.0 eth0: link down
Press the [f] key and hit [enter] to enter failsafe mode
Press the [1], [2], [3] or [4] key and hit [enter] to select the debug level
[    6.035670] mv643xx_eth_port mv643xx_eth_port.0 eth0: link up, 1000 Mb/s, full duplex, flow control disabled
[    6.045590] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[    7.015095] procd: - early -
[    7.018452] procd: - watchdog -
[    7.550163] procd: - watchdog -
[    7.553516] procd: - ubus -
[    7.608534] procd: - init -
Please press Enter to activate this console.
[    7.902411] kmodloader: loading kernel modules from /etc/modules.d/*
[    8.020590] urngd: v1.0.2 started.
[    8.045486] Loading modules backported from Linux version v5.10.42-0-g65859eca4dff
[    8.053126] Backport generated by backports.git v5.10.42-1-0-gbee5c545
[    8.137968] xt_time: kernel timezone is -0000
[    8.205984] random: crng init done
[    8.209432] random: 7 urandom warning(s) missed due to ratelimiting
[    8.268228] PPP generic driver version 2.4.2
[    8.275628] NET: Registered protocol family 24
[    8.352745] kmodloader: done loading kernel modules from /etc/modules.d/*
[   21.125504] Marvell 88E1318S f1072004.mdio-bus-mii:00: marvell,reg-init!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
[   21.151996] br-lan: port 1(eth0) entered blocking state
[   21.157277] br-lan: port 1(eth0) entered disabled state
[   21.162705] device eth0 entered promiscuous mode
[   21.892638] device eth0 left promiscuous mode
[   21.897100] br-lan: port 1(eth0) entered disabled state
[   26.051330] reboot: Power down
[   26.055015] reboot: Restarting system with command 'Power off'
```

Leo-PL commented:
Something is broken quite badly here. I still have my build of backport based on kernel 4.14.229, which did work, and new, on 4.14.235 which doesn't. There are around 500 commits between those tags. I've skimmed through them, but nothing interesting came up. Maybe before, the actual order of probing of PHY driver vs gpio-button-hotplug was different? I wonder if adding debouncing to the buttons, with time substantially longer than the low pulse duration caused by PHY reconfiguration could solve the issue - however it would still be workaround.

I'm thinking how to automate bisection of the kernel tree, together with building and running the firmware on target - at least I know I have a hardware setup capable of doing that.

CHKDSK88 commented:
If I use `phytool` in userspace,  it's ok. But if I set the same registers via dts, it's a lottery,  

Maybe @natisbad have some idea?

Leo-PL commented:
> 
> 
> @Leo-PL
> 
> > Also, freshly build initramfs images for some reason, seem to crash very early on, even before the initial "OK" is printed. Can they be too big? With current master, initramfs-uImage weighs in 12337598 bytes for me.
> 
> Initramfs image must be small. I'm not sure what size of image is maximum for proper loading. I had the same problem when I test images prepared by @obsy from eko.one.pl.
> 
> > I just rebased this PR together with my patchset, and for some strange reason, when OpenWrt boots, at the time around link establishes, a poweroff is triggered by kernel, and system turns off. This did not happen during my previous tests, and really nothing has changed in this PR, so I don't suspect it to have the problem.
> 
> I have the same problem.
> 
> > One thought came to my mind. I don't have the schematics for the device - but isn't WoL working by asserting the "power button" line internally?
> 
> Exactly, it looks that is problem with LED2/INTn pin connected to power button. It could pull down power button.
> I pushed partial solution of powering-off problem. Now sometimes work perfect, sometimes not. It depends, how fast phy will be configured. If gpio-hotplug is faster, then it reboot. If phy are faster, it works.
> 
> The only way is to delay start of gpio-hotplug or do some changes in phy configuration. Do You have any idea how to make it better?

Another idea: maybe we can configure this pin from U-boot somehow?
Also, why the note about upstream `gpio-poweroff` was dropped from the device tree?

CHKDSK88 commented:
@Leo-PL 
 
> Another idea: maybe we can configure this pin from U-boot somehow? We have a "phyWrite" command available there.

I'm on holiday till next Sunday, so I will look again after 16.08.21. Maybe this will work.

> Also, why the note about upstream `gpio-poweroff` was dropped from the device tree?

I found  why author do it:
https://github.com/torvalds/linux/commit/bfc2e5f560c7af89ee1cece8a0018a44011c24b2

I tried multiple possible solutions. I removed it during tests. But if we use my patch to poweroff driver, this entry is useless.

> Edit: I'll try this in spare time, which I don't have much today, and try bisecting the kernel using "external kernel tree" feature of build system and some automation.

Did You try Debian? I wonder, if Openwrt's netifd isn't a problem.

Leo-PL commented:
> 
> 
> @Leo-PL
> 
> > Another idea: maybe we can configure this pin from U-boot somehow? We have a "phyWrite" command available there.
> 
> I'm on holiday till next Sunday, so I will look again after 16.08.21. Maybe this will work.
It's likely, that I'll have time to play with this earliest next weekend as well. 
> 
> > Also, why the note about upstream `gpio-poweroff` was dropped from the device tree?
> 
> I found why author do it:
> [torvalds/linux@bfc2e5f](https://github.com/torvalds/linux/commit/bfc2e5f560c7af89ee1cece8a0018a44011c24b2)
> 
> I tried multiple possible solutions. I removed it during tests of workaround. But if we use my patch to poweroff driver, this entry is useless.

Acked, makes sense.
> 
> > Edit: I'll try this in spare time, which I don't have much today, and try bisecting the kernel using "external kernel tree" feature of build system and some automation.
> 
> Did You try Debian? I wonder, if Openwrt's netifd isn't a problem.

Not yet, but I wouldn't suspect netifd here. I'd be happy to run Debian on this box as well, but OpenWrt is my go-to embedded framework.

CHKDSK88 commented:
In the meantime my poweroff patches were accepted:
https://lkml.org/lkml/2021/8/13/773

Leo-PL commented:
> In the meantime my poweroff patches were accepted:
> https://lkml.org/lkml/2021/8/13/773

Congratulations!

Leo-PL commented:
@CHKDSK88 I ran a bisect session on this commit, and narrowed the problem down to this commit, actually in OpenWrt, not in kernel: 467cd378db3359814866ec4121fdb4417783241e (base-files: failsafe: Fix IP configuration) - How strange indeed it is.
Reverting it on top of of master did help.
To perform bisection, I did the following:
1. Rebase this patchset on top of b72157984268b6542dd667859e4aed9cc2bbb7f5. Checkout as `bisect_good`
2. Remove the partial workaround in device tree with `marvell,phy-init` property
3. Rebase b30924873062be47250e6b872347461c8221688d on top of `bisect_good`. Checkout as `bisect bad`
4. Set the unit to autoboot from TFTP from my buildserver using `setenv bootcmd 'setenv ipaddr <DEVICE_IP>; setenv serverip <BUILDSERVER_IP>; tftpboot 0x1200000 openwrt-uImage; bootm 0x1200000'; saveenv; reset`
5. Prepare a symlink to build output from OpenWrt repository: `cd /srv/tftp/; ln -s </path/to/build/dir>/bin/targets/kirkwood/generic/openwrt-kirkwood-netgear_readynas-duo-v2-initramfs-uImage openwrt-uImage`
6. Create a config seed for the build process as `mydiffconfig`:
```
CONFIG_TARGET_kirkwood=y
CONFIG_TARGET_kirkwood_DEVICE_netgear_readynas-duo-v2=y
CONFIG_TARGET_BOARD="kirkwood"
CONFIG_DEVEL=y
CONFIG_COLLECT_KERNEL_DEBUG=y
CONFIG_IMAGEOPT=y
CONFIG_PACKAGE_kmod-ata-core=m
CONFIG_PACKAGE_kmod-ata-marvell-sata=m
CONFIG_PACKAGE_kmod-crypto-crc32c=m
CONFIG_PACKAGE_kmod-crypto-hash=m
CONFIG_PACKAGE_kmod-fs-ext4=m
CONFIG_PACKAGE_kmod-hwmon-core=m
CONFIG_PACKAGE_kmod-hwmon-g762=m
CONFIG_PACKAGE_kmod-i2c-core=m
CONFIG_PACKAGE_kmod-lib-crc16=m
CONFIG_PACKAGE_kmod-rtc-rs5c372a=m
CONFIG_PACKAGE_kmod-scsi-core=m
CONFIG_PACKAGE_kmod-usb-xhci-hcd=m
CONFIG_PACKAGE_kmod-usb3=m

```
7. Add the following snippet to ~/.ssh/config (I had this already) so we can connect to TFTP-booted image without checking host keys, etc, through SSH:
```
Host openwrt-recovery
        Hostname 192.168.1.1
        User root
        UserKnownHostsFile /dev/null
        StrictHostKeyChecking no
        CheckHostIP no

```

8. Create test script which builds OpenWrt using this, then boots the unit through WOL and TFTP; and verifies if the issue occurs, as `do_all.sh`:
```
#!/bin/bash
cp mydiffconfig .config
make defconfig
time make -j8 || exit 125
wakeonlan -i 10.0.0.255 2C:B0:5D:BE:EF:DA
sleep 300
ping -c 4 netgear-rnd2000 || exit 1
ssh -o Hostname=<NAS_HOSTNAME_OR_IP> openwrt-recovery poweroff
exit 0
```
9. Perform actual bisect, by executing this
```
git bisect start
git bisect good bisect_good
git bisect bad bisect_bad
git bisect run ./do_all.sh
```

10. ...
11. Profit! The result was this: 
```
git bisect start
# good: [c7b79b72cc57f82552fafa0afbcf1c1f3ffa9531] Disable poweroff workaround just to be sure
git bisect good c7b79b72cc57f82552fafa0afbcf1c1f3ffa9531
# bad: [6b6550e76d7a7fdd98ae4b2d0a0b21015c978706] generic: add mac-address property for NVMEM mac addresses
git bisect bad 6b6550e76d7a7fdd98ae4b2d0a0b21015c978706
# bad: [08b4c0a2ea2ca0ae620c0c97d2add96c7aaf6a5b] Revert "dnsmasq: Update to version 2.86test3"
git bisect bad 08b4c0a2ea2ca0ae620c0c97d2add96c7aaf6a5b
# good: [c282de3e7082b4c19d319f4b77675e8ad4a12898] tools/squashfskit4: fix compilation under big endian
git bisect good c282de3e7082b4c19d319f4b77675e8ad4a12898
# bad: [867804645cfe5acec600f01e579d94dc6dbdf34a] ramips: mt7620: remove useless GMAC nodes
git bisect bad 867804645cfe5acec600f01e579d94dc6dbdf34a
# good: [7bf48240aa42c6024fee207689d062e9a9c33a9f] hostapd: make country3 option configurable
git bisect good 7bf48240aa42c6024fee207689d062e9a9c33a9f
# good: [b345fc3107ba41f419214da4212577954b2b7372] kernel: bump 5.10 to 5.10.44
git bisect good b345fc3107ba41f419214da4212577954b2b7372
# bad: [680edeacdf41f5faaeddc2174967285c8e6b26fe] base-files: failsafe: Remove the VLAN modifier from interface name
git bisect bad 680edeacdf41f5faaeddc2174967285c8e6b26fe
# good: [5dde23fb2a9e381659b575696a1588aca36a6886] kernel: Backport patch to automatically bring up DSA master when opening user port
git bisect good 5dde23fb2a9e381659b575696a1588aca36a6886
# bad: [04f42b28961402199526ade4e1ac9f57ea5c8c47] base-files: failsafe: Fix IP configuration
git bisect bad 04f42b28961402199526ade4e1ac9f57ea5c8c47
# first bad commit: [04f42b28961402199526ade4e1ac9f57ea5c8c47] base-files: failsafe: Fix IP configuration
```
And then reverting last commit on my local tree did indeed help!

Now, the question remains, why does this commit affect the power-up sequence. 

OT: Nice thing is, that I haven't actually need to use my smart plug running https://github.com/arendst/Tasmota to perform power-cycles on the unit, but still it is a very useful tool in an embedded CI environment.

Leo-PL commented:
I ran another bisect session, on full image including my patches, and using sysupgrade in the process.
And that's, what came out:
```
git bisect start
# good: [cc923c6cd98dc0c072c1ed285654a2bfb09f878c] Disable poweroff workaround just to be sure
git bisect good cc923c6cd98dc0c072c1ed285654a2bfb09f878c
# bad: [3e9e510d14037a3d83587077b8b64f01c0ef2adf] generic: add mac-address property for NVMEM mac addresses
git bisect bad 3e9e510d14037a3d83587077b8b64f01c0ef2adf
# bad: [1e0ad195449771840ad6bcc9c5b2c40ba29b2569] Revert "dnsmasq: Update to version 2.86test3"
git bisect bad 1e0ad195449771840ad6bcc9c5b2c40ba29b2569
# good: [fda26dcbaa588a45154926351dc8a73c97249c7a] tools/squashfskit4: fix compilation under big endian
git bisect good fda26dcbaa588a45154926351dc8a73c97249c7a
# bad: [49bf7fb1f3afa5b41f1bf8261c1743e11f95dfe0] ramips: mt7620: remove useless GMAC nodes
git bisect bad 49bf7fb1f3afa5b41f1bf8261c1743e11f95dfe0
# good: [43873ff7d25637ff6fad636405b0c4ff6fdc6298] hostapd: make country3 option configurable
git bisect good 43873ff7d25637ff6fad636405b0c4ff6fdc6298
# bad: [e300ee7a0b3310b40ad5347cc80967db34f080ce] kernel: bump 5.10 to 5.10.44
git bisect bad e300ee7a0b3310b40ad5347cc80967db34f080ce
# good: [90701978b8a7446dacec0594e51549a1ac126d7f] base-files: fix enabled for services with only STOP
git bisect good 90701978b8a7446dacec0594e51549a1ac126d7f
# good: [39e9e2278e5df1142fcb1de946f5060cfb84e803] realtek: Fix buffer length calculation on RTL8380 with CRC offload
git bisect good 39e9e2278e5df1142fcb1de946f5060cfb84e803
# good: [9d5cfb79c9de08768d4449f27a223b7714bf85ee] kernel: crypto: limit crypto-hw-hifn-795x to devices with pci support
git bisect good 9d5cfb79c9de08768d4449f27a223b7714bf85ee
# first bad commit: [e300ee7a0b3310b40ad5347cc80967db34f080ce] kernel: bump 5.10 to 5.10.44
```
I am puzzled, because was running kernel 5.4 in my tests. 
There are a few changes, that I had to do:

1. Keep a clean initramfs image, to boot the unit through TFTP, modified to include a custom boot command:
`fw_setenv bootcmd "nand read.e 0x1200000 0x200000 0x600000; setenv bootcmd 'run bootcmd_tftp'; saveenv; bootm 0x1200000"` in /target/linux/kirkwood/base-files/lib/upgrade/platform.sh, where `bootcmd_tftp` was the boot command for TFTP build normally used for initial installation. This causes the unit to revert to TFTP boot, once sysupgraded image is launched. 
```
diff --git a/target/linux/kirkwood/base-files/lib/upgrade/platform.sh b/target/linux/kirkwood/base-files/lib/upgrade/platform.sh
index 5c5415baf2..f9267be8c9 100644
--- a/target/linux/kirkwood/base-files/lib/upgrade/platform.sh
+++ b/target/linux/kirkwood/base-files/lib/upgrade/platform.sh
@@ -33,7 +33,7 @@ platform_do_upgrade() {
                        mkdir -p /var/lock
                        touch /var/lock/fw_printenv.lock
                        fw_setenv bootargs 'console=ttyS0,115200n8 earlyprintk'
-                       fw_setenv bootcmd 'nand read.e 0x1200000 0x200000 0x600000;bootm 0x1200000'
+                       fw_setenv bootcmd "nand read.e 0x1200000 0x200000 0x600000; setenv bootcmd 'run bootcmd_tftp'; saveenv; bootm 0x1200000"
                fi
                nand_do_upgrade "$1"
                ;;
```
There is no need to include this in test images, as this variable is always set by image booted from TFTP, not the actual test image.

2.  Use slightly modified script to be executed by `git bisect`:
```
#!/bin/bash
cp mydiffconfig .config
make defconfig
time make -j8 || exit 125
wakeonlan -i 10.0.0.255 2C:B0:5D:BE:EF:DAsquashfs-sysupgrade.bin'
sleep 240

#!/bin/bash
cp mydiffconfig .config
make defconfig
time make -j<NUMBER_CPUS> || exit 125
wakeonlan -i <SUBNET_BROADCAST_ADDR <DEVICE_MAC>
sleep 240
scp -o Hostname=<NAS_IP_OR_HOSTNAME> bin/targets/kirkwood/generic/openwrt-kirkwood-netgear_readynas-duo-v2-squashfs-sysupgrade.bin openwrt-recovery:/tmp/
ssh -o Hostname=<NAS_IP_OR_HOSTNAME> openwrt-recovery 'sysupgrade -n /tmp/openwrt-kirkwood-netgear_readynas-duo-v2-squashfs-sysupgrade.bin'
sleep 240
ping -c 4 <NAS_HOSTNAME_OR_IP> || exit 1 #The actual test happens here
ssh -o Hostname=<NAS_HOSTNAME_OR_IP> openwrt-recovery poweroff #Prepare for next test cycle
exit 0
```

skoczko commented:
Guys, what's blocking this from getting merged? Did people involved lose interest?

Leo-PL commented:
@skoczko We haven't figured out a way to consistently fix the power button issue mis-triggered by PHY configuration.

emesio commented:
Thats a pity. Any chances for 'unofficial' build (bins)? If the power button was the only issue - then resetting it from smart plug is not a problem, and the box looks perfect for simple openwrt nas... :)

CHKDSK88 commented:
It could be merged if any coredev accept, that power button will not close system. That's only one problem with that device...

If @adschm or @mans0n accept that temporary solution, I could prepare new version.

Leo-PL commented:
I agree that it's better than nothing - one could still use "poweroff" command.
@CHKDSK88 One idea struck me now - could we use a very long debounce time, like 1 second, on that line as a workaround?

CHKDSK88 commented:
@Leo-PL 

I will check it. But if it will not help, I will prepare final version without power button functionality.

CHKDSK88 commented:
@Leo-PL 
It didn't help. I will prepare version without support for power button soon.

Leo-PL commented:
@CHKDSK88, +1, makes total sense. Please ping me when you push update, I'll rebuild and retest.

CHKDSK88 commented:
@Leo-PL 

It's ready for testing. I hope, that's is ready for merge too.

@mans0n 
Could You look at this PR in free time?

Leo-PL commented:
Tested-by: Lech Perczak <lech.perczak@gmail.com>

I noticed, that the power button inactivates, when the link is finally up after setting up interfaces. Unit properly powers off after a "poweroff" command. I'll try to look into the PHY driver in near future.

CHKDSK88 commented:
@Leo-PL 
Thanks for the help.
I can't force PHY to keep INTn pin state when PHY is resetting. I hope, that workaround is enough for merge.

Leo-PL commented:
Besides, do you have a datasheet for this PHY, by any chance?

CHKDSK88 commented:
> Besides, do you have a datasheet for this PHY, by any chance?

You can't find it in the internet. :/

mans0n commented on target/linux/kirkwood/base-files/lib/upgrade/platform.sh:
```diff
@@ -16,6 +27,16 @@ platform_do_upgrade() {
+			mkdir -p /var/lock
+			touch /var/lock/fw_printenv.lock
+			fw_setenv bootargs 'console=ttyS0,115200n8 earlyprintk'
+			fw_setenv bootcmd 'nand read.e 0x1200000 0x200000 0x600000;bootm 0x1200000'
```
It's still lingering in my mind. I think this can be done manually once during the installation. We should access u-boot shell for installation anyway, why should we split the process?
I mean, `setenv bootcmd ... && saveenv` could be inserted somewhere around `bootm 0x1200000` in the commit message.

CHKDSK88 replied:
Ok. I removed it.

mans0n replied:
And then we don't need `platform_check_image()` change anymore, right?

CHKDSK88 replied:
I was too fast. Now should be ok.

mans0n reviewed (commented):
Sorry for the delay.
It mostly seems good, just one nitpick:

chunkeey commented:
Merged. Thanks!
